File: src/foam/dao/FilteredDAO.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 * http://www.apache.org/licenses/LICENSE-2.0
 */

package foam.dao;

import foam.core.X;
import foam.mlang.predicate.Predicate;
import foam.mlang.order.Comparator;

public class FilteredDAO
  extends ProxyDAO
{
  protected Predicate predicate_;

  public FilteredDAO(Predicate predicate, DAO delegate) {
    predicate_ = predicate;
    setDelegate(delegate);
    if ( delegate instanceof ProxyDAO ) {
      setX(((ProxyDAO)delegate).getX());
    }
  }

  public FilteredDAO(X x, Predicate predicate, DAO delegate) {
    predicate_ = predicate;
    setX(x);
    setDelegate(delegate);
  }

  public FilteredDAO() {}

  public void setPredicate(Predicate predicate) {
    predicate_ = predicate;
  }

  private Predicate getPredicate(Predicate arg) {
    return arg == null ? predicate_ : foam.mlang.MLang.AND(predicate_, arg);
  }

  @Override
  public Sink select_(X x, Sink s, long skip, long limit, Comparator order, Predicate predicate) {
    return super.select_(x, s, skip, limit, order, getPredicate(predicate));
  }

  @Override
  public void removeAll_(X x, long skip, long limit, Comparator order, Predicate predicate) {
    super.removeAll_(x, skip, limit, order, getPredicate(predicate));
  }

  @Override
  public void listen_(X x, Sink sink, Predicate predicate) {
    super.listen_(x, sink, getPredicate(predicate));
  }
}

File: src/foam/dao/FilteredDAO.js
foam.CLASS({
  package: 'foam.dao',
  name: 'FilteredDAO',
  extends: 'foam.dao.ProxyDAO',

  requires: [
    'foam.mlang.predicate.And'
  ],

  properties: [
    {
      // TODO: FObjectProperty of Predicate. Doing this currently breaks java.
      class: 'FObjectProperty',
      of: 'foam.mlang.predicate.Predicate',
      swiftType: 'foam_mlang_predicate_Predicate',
      name: 'predicate'
    }
  ],

  methods: [
    {
      name: 'find_',
      code: function find_(x, key) {
        var predicate = this.predicate;
        return this.delegate.find_(x, key).then(function(o) {
          return predicate.f(o) ? o : null;
        });
      },
      javaCode: `foam.core.FObject ret = super.find_(x, id);
if ( ret != null && getPredicate().f(ret) ) return ret;
return null;`
    },

    
    {
      name: 'select_',
      code: function(x, sink, skip, limit, order, predicate) {
        return this.delegate.select_(
          x, sink, skip, limit, order,
          predicate ?
            this.And.create({ args: [this.predicate, predicate] }) :
            this.predicate);
      },
      swiftCode: function() {/*
return try delegate.select_(
  x, sink, skip, limit, order,
  predicate != nil ?
    And_create(["args": [self.predicate, predicate!] ]) :
    self.predicate)
                             */},
      javaCode: 'return super.select_(x, sink, skip, limit, order, predicate == null ? getPredicate() : foam.mlang.MLang.AND(getPredicate(), predicate));'
    },

    {
      name: 'removeAll_',
      code: function removeAll_(x, skip, limit, order, predicate) {
        return this.delegate.removeAll_(
          x, skip, limit, order,
          predicate ?
            this.And.create({ args: [this.predicate, predicate] }) :
          this.predicate);
      },
      javaCode: 'super.removeAll_(x, skip, limit, order, predicate == null ? getPredicate() : foam.mlang.MLang.AND(getPredicate(), predicate));'
    },

    {
      name: 'listen_',
      code: function listen_(x, sink, predicate) {
        return this.delegate.listen_(
          x, sink,
          predicate ?
            this.And.create({ args: [this.predicate, predicate] }) :
            this.predicate);
      },
      swiftCode: `
return try delegate.listen_(
  x, sink,
  predicate != nil ?
    And_create(["args": [self.predicate, predicate]]) :
    predicate)
      `,
      javaCode: 'super.listen_(x, sink, predicate == null ? getPredicate() : foam.mlang.MLang.AND(getPredicate(), predicate));'
    },
  ]
});
